Teach salt.utils.minions.CkMinions.connected_ids() about IPv6

dwfreed opened the pull request:
### What does this PR do?

This teaches salt.utils.minions.CkMinions.connected_ids() about IPv6, so that IPv6 minions show up in the manage.list_state and family runners, as well as presence events, when the master runs on a system with /proc/net/tcp{,6}.

### What issues does this PR fix or reference?

While I was in this section of code, I fixed some logic issues around including any localhost minion (including IPv6 localhost minion), while avoiding including any minion that isn't connected.  This allows any localhost minion to always show up in presence events.

I also fixed salt.utils.network.local_port_tcp() and salt.utils.network.active_tcp() so they don't include listening ports when /proc/net/tcp is available, to match their OS-specific implementations, and made it so IPv6 connections don't overwrite IPv4 connections in active_tcp() results.

### Previous Behavior

IPv6 minions don't show up in manage.list_state and family runner results or presence events on masters with /proc/net/tcp{,6}.  salt.utils.network.local_port_tcp() and salt.utils.network.active_tcp() include listening ports when /proc/net/tcp{,6} is available.  salt.utils.network.active_tcp() is missing some IPv4 connections when there are IPv6 connections.

### New Behavior

IPv6 minions are included in manage.list_state and family runner results and presence events on masters with /proc/net/tcp{,6}.  salt.utils.network.local_port_tcp() and salt.utils.network.active_tcp() don't include listening ports when /proc/net/tcp{,6} is available.  salt.utils.network.active_tcp() includes all IPv4 and IPv6 connections.

### Tests written?

No

### Commits signed with GPG?

Yes

### Miscellaneous notes

I opened this PR against develop because I changed the API of CkMinions.connected_ids() as the include_localhost parameter was no longer needed.  I would like to see this backported to at least 2018.3, though, and preferably make it in before 2018.3.2 is released, so I don't have to continue to apply this fix locally.  If changing the API of this function in a stable release is an issue, then the function could still accept the parameter but just not do anything with it.

I'm still not happy with how CkMinions.connected_ids() is implemented, because there's one presumably common case and a few edge cases that are not handled:

* NATed minions won't be listed (presumably common case)
* If there are multiple minions with the same IP address, they will all be listed if at least one is connected from that address
* If there are multiple masters listening on the same port but different addresses, minions can be listed that aren't connected to the master being interrogated
* (possibly Linux-specific) If a localhost minion connects to the master using an address in 127.0.0.0/8 that either is not explicitly configured on the loopback interface or the master is not explicitly listening on, it won't be listed

Fixing any of these would require a lot more work.

isbm commented on salt/utils/minions.py:
```diff
@@ -643,16 +647,19 @@ def connected_ids(self, subset=None, show_ipv4=False, include_localhost=False):
                         break
+                for ipv6 in grains.get('ipv6', []):
+                    if ipv6 in addrs:
+                        if show_ipv4:
```
This is request to show IPv4, and still you putting IPv6. I guess, this parameter needs either be renamed to "show_ip" (preferred) or we need to add another one (no idea why). In fact, this parameter/option is only used once, in manage runner, list_state function.

gtmanfred reviewed (changes requested):
Can the localhost changes be split out to a different PR so that we can make sure this gets reviewed correctly?

I just want to make sure something isn't missed while localhost is being changed.

Thanks,
Daniel

gtmanfred commented:
Ok, then can you split it into two commits, the first one fixing the localhost stuff for ipv4, and then the second one adding ipv6 support so that we can see the difference?

gtmanfred commented:
Can you revert the changes to `doc/man/salt.7`? these are automatically generated when packaged, they do not need to be modified.

gtmanfred commented on salt/runners/manage.py:
```diff
@@ -201,7 +201,7 @@ def up(tgt='*', tgt_type='glob', timeout=None, gather_job_timeout=None):  # pyli
 
 
-def list_state(subset=None, show_ipv4=False, state=None):
+def list_state(subset=None, show_ip=False, state=None):
```
This change will need to go through a deprecation process.

https://docs.saltstack.com/en/latest/topics/development/deprecations.html

gtmanfred replied:
Where you have show_ipv4, and show_ip for 2 releases, and then it can be removed.

dwfreed committed:
utils/network: only include established connections when reading /proc

This matches the behavior of the OS-specific implementations used when
/proc is not available. Remove the check for 0.0.0.0 in
CkMinions.connected_ids since it won't come through anymore.

dwfreed committed:
utils/network: fix IPv6 connections overwriting IPv4 in active_tcp()

dwfreed committed:
utils/minions: fix localhost logic in CkMinions.connected_ids

We don't need loopback addresses in the address list to find any
localhost minion.

dwfreed committed:
utils/minions: Deprecate include_localhost param of CkMinions.connected_ids

This is no longer used.

dwfreed committed:
utils/network: handle v4-mapped v6 in hex2ip

Add some more tests to cover it.

dwfreed committed:
utils/minions: teach CkMinions.connected_ids IPv6

This allows things like `salt-run manage.present` and the presence
system to see IPv6 minions.

dwfreed committed:
utils/minions: rename show_ipv4 param of CkMinions.connected_ids

Also update the manage runner functions that used it, as well as the
documentation.

dwfreed committed:
Update presence_events docs to remove localhost comment

Localhost minions are now included in presence events.
